File: src/worldpainterStubs/java/World2.java
package org.pepsoft.worldpainter;

import org.pepsoft.minecraft.Direction;
import org.pepsoft.util.AttributeKey;
import org.pepsoft.util.ProgressReceiver;
import org.pepsoft.worldpainter.exporting.WorldExportSettings;
import org.pepsoft.worldpainter.history.HistoryEntry;
import org.pepsoft.worldpainter.layers.Layer;

import java.awt.*;
import java.beans.PropertyChangeListener;
import java.io.File;
import java.io.IOException;
import java.io.Serializable;
import java.util.*;
import java.util.List;
import java.util.zip.ZipOutputStream;

public interface World2Interface extends Serializable, Cloneable {
    @Deprecated
    int BIOME_ALGORITHM_NONE = -1;
    @Deprecated
    int BIOME_ALGORITHM_CUSTOM_BIOMES = 6;
    @Deprecated
    int BIOME_ALGORITHM_AUTO_BIOMES = 7;
    int DEFAULT_MAX_HEIGHT = org.pepsoft.minecraft.Constants.DEFAULT_MAX_HEIGHT_ANVIL;
    long DEFAULT_OCEAN_SEED = 27594263L; // A seed with a large ocean around the origin, and not many mushroom islands nearby. Should be used with Large Biomes long DEFAULT_LAND_SEED = 227290L; // A seed with a huge continent around the origin. Should be used with Large Biomes
    /**
     * A {@link String} containing the WorldPainter version with which this file
     * was saved.
     */
    String METADATA_KEY_WP_VERSION = "org.pepsoft.worldpainter.wp.version";
    /**
     * A {@link String} containing the WorldPainter build with which this file
     * was saved.
     */
    String METADATA_KEY_WP_BUILD = "org.pepsoft.worldpainter.wp.build";
    /**
     * A {@link Date} containing the time at which this file was saved.
     */
    String METADATA_KEY_TIMESTAMP = "org.pepsoft.worldpainter.timestamp";
    /**
     * An optional two dimensional {@link String}[][2] array containing the
     * plugins installed in the WorldPainter instance which saved the file. One
     * row per plugin, which each row containing two elements, the first being
     * the plugin name, the second being the plugin version.
     *
     * <p>May be {@code null} if no non-standard plugins were present.
     */
    String METADATA_KEY_PLUGINS = "org.pepsoft.worldpainter.plugins";
    /**
     * A string containing the name of the world.
     */
    String METADATA_KEY_NAME = "name";

    long getChangeNo();

    String getName();

    void setName(String name);

    boolean isCreateGoodiesChest();

    void setCreateGoodiesChest(boolean createGoodiesChest);

    Point getTileCoordinates(int worldX, int worldY);

    Point getTileCoordinates(Point worldCoords);

    Point getSpawnPoint();

    void setSpawnPoint(Point spawnPoint);

    File getImportedFrom();

    void setImportedFrom(File importedFrom);

    boolean isMapFeatures();

    void setMapFeatures(boolean mapFeatures);

    GameType getGameType();

    void setGameType(GameType gameType);

    void addPropertyChangeListener(PropertyChangeListener listener);

    void addPropertyChangeListener(String propertyName, PropertyChangeListener listener);

    void removePropertyChangeListener(PropertyChangeListener listener);

    void removePropertyChangeListener(String propertyName, PropertyChangeListener listener);

    @Deprecated
    Dimension getDimension(int dim);

    boolean isDimensionPresent(Anchor anchor);

    Dimension getDimension(Anchor anchor);

    Set<Dimension> getDimensions();

    Set<Dimension> getDimensionsWithRole(Dimension.Role role, boolean inverted, int id);

    void addDimension(Dimension dimension);

    Dimension removeDimension(Anchor anchor);

    MixedMaterial getMixedMaterial(int index);

    void setMixedMaterial(int index, MixedMaterial material);

    int getMinHeight();

    void setMinHeight(int minHeight);

    int getMaxHeight();

    void setMaxHeight(int maxHeight);

    @Deprecated
    Generator getGenerator();

    Platform getPlatform();

    void setPlatform(Platform platform);

    boolean isAskToConvertToAnvil();

    void setAskToConvertToAnvil(boolean askToConvertToAnvil);

    boolean isAskToRotate();

    void setAskToRotate(boolean askToRotate);

    Direction getUpIs();

    void setUpIs(Direction upIs);

    boolean isAllowMerging();

    void setAllowMerging(boolean allowMerging);

    boolean isAllowCheats();

    void setAllowCheats(boolean allowCheats);

    @Deprecated
    String getGeneratorOptions();

    boolean isExtendedBlockIds();

    void setExtendedBlockIds(boolean extendedBlockIds);

    int getDifficulty();

    void setDifficulty(int difficulty);

    WorldExportSettings getExportSettings();

    void setExportSettings(WorldExportSettings exportSettings);

    List<HistoryEntry> getHistory();

    void addHistoryEntry(int key, Serializable... args);

    Map<String, Object> getMetadata();

    void setMetadata(Map<String, Object> metadata);

    BorderSettings getBorderSettings();

    File getMergedWith();

    void setMergedWith(File mergedWith);

    List<File> getDataPacks();

    void setDataPacks(List<File> dataPacks);

    <T> Optional<T> getAttribute(AttributeKey<T> key);

    <T> void setAttribute(AttributeKey<T> key, T value);

    void transform(CoordinateTransform transform, ProgressReceiver progressReceiver) throws ProgressReceiver.OperationCancelled;

    void transform(Anchor anchor, CoordinateTransform transform, ProgressReceiver progressReceiver) throws ProgressReceiver.OperationCancelled;

    void clearLayerData(Layer layer);

    long measureSize();

    void save(ZipOutputStream out) throws IOException;

    public enum Warning {
        /**
         * Warn the user that automatic biomes are now the default and are enabled.
         */
        AUTO_BIOMES_ENABLED,

        /**
         * Warn the user that automatic biomes were previously in use but are now disabled.
         */
        AUTO_BIOMES_DISABLED,

        /**
         * Warn the user that one or more custom terrain types were missing and have been replaced with magenta wool.
         */
        MISSING_CUSTOM_TERRAINS,

        /**
         * Warn the user that the Superflat settings could not be parsed and were reset to defaults.
         */
        SUPERFLAT_SETTINGS_RESET,

        /**
         * The game type was lost and was reset to Survival.
         */
        GAME_TYPE_RESET
    }
}

File: src/worldpainterStubs/typescript/World2.ts
import {Dimension} from "./Dimension";

type Anchor = any; // Define the actual type for Anchor
type Generator = any; // Define the actual type for Generator
type Platform = any; // Define the actual type for Platform
type GameType = any; // Define the actual type for GameType
type WorldExportSettings = any; // Define the actual type for WorldExportSettings
type HistoryEntry = any; // Define the actual type for HistoryEntry
type BorderSettings = any; // Define the actual type for BorderSettings
type AttributeKey<T> = any; // Define the actual type for AttributeKey
type CoordinateTransform = any; // Define the actual type for CoordinateTransform
type ProgressReceiver = any; // Define the actual type for ProgressReceiver
type Layer = any; // Define the actual type for Layer
type Optional<T> = any; // Define the actual type for Optional
type Direction = any;
type Role = any;

type Set<T> = any
type Map<A,B> = any
interface Point {
    x: number;
    y: number;
}


type PropertyChangeListener = any;
type MixedMaterial = any;
type Serializable = any;
type ZipOutputStream = any;

export interface World2 {
    readonly BIOME_ALGORITHM_NONE: number;
    readonly BIOME_ALGORITHM_CUSTOM_BIOMES: number;
    readonly BIOME_ALGORITHM_AUTO_BIOMES: number;
    readonly DEFAULT_MAX_HEIGHT: number;
    readonly DEFAULT_OCEAN_SEED: number;
    readonly DEFAULT_LAND_SEED: number;
    readonly METADATA_KEY_WP_VERSION: string;
    readonly METADATA_KEY_WP_BUILD: string;
    readonly METADATA_KEY_TIMESTAMP: string;
    readonly METADATA_KEY_PLUGINS: string;
    readonly METADATA_KEY_NAME: string;

    getChangeNo(): number;

    getName(): string;

    setName(name: string): void;

    isCreateGoodiesChest(): boolean;

    setCreateGoodiesChest(createGoodiesChest: boolean): void;

    getTileCoordinates(worldX: number, worldY: number): Point;

    getTileCoordinates(worldCoords: Point): Point;

    getSpawnPoint(): Point;

    setSpawnPoint(spawnPoint: Point): void;

    getImportedFrom(): File;

    setImportedFrom(importedFrom: File): void;

    isMapFeatures(): boolean;

    setMapFeatures(mapFeatures: boolean): void;

    getGameType(): GameType;

    setGameType(gameType: GameType): void;

    addPropertyChangeListener(listener: PropertyChangeListener): void;

    addPropertyChangeListener(propertyName: string, listener: PropertyChangeListener): void;

    removePropertyChangeListener(listener: PropertyChangeListener): void;

    removePropertyChangeListener(propertyName: string, listener: PropertyChangeListener): void;

    getDimension(dim: number): Dimension; // Deprecated
    isDimensionPresent(anchor: Anchor): boolean;

    getDimension(anchor: Anchor): Dimension;

    getDimensions(): Set<Dimension>;

    getDimensionsWithRole(role: Role, inverted: boolean, id: number): Set<Dimension>;

    addDimension(dimension: Dimension): void;

    removeDimension(anchor: Anchor): Dimension;

    getMixedMaterial(index: number): MixedMaterial;

    setMixedMaterial(index: number, material: MixedMaterial): void;

    getMinHeight(): number;

    setMinHeight(minHeight: number): void;

    getMaxHeight(): number;

    setMaxHeight(maxHeight: number): void;

    getGenerator(): Generator; // Deprecated
    getPlatform(): Platform;

    setPlatform(platform: Platform): void;

    isAskToConvertToAnvil(): boolean;

    setAskToConvertToAnvil(askToConvertToAnvil: boolean): void;

    isAskToRotate(): boolean;

    setAskToRotate(askToRotate: boolean): void;

    getUpIs(): Direction;

    setUpIs(upIs: Direction): void;

    isAllowMerging(): boolean;

    setAllowMerging(allowMerging: boolean): void;

    isAllowCheats(): boolean;

    setAllowCheats(allowCheats: boolean): void;

    getGeneratorOptions(): string; // Deprecated
    isExtendedBlockIds(): boolean;

    setExtendedBlockIds(extendedBlockIds: boolean): void;

    getDifficulty(): number;

    setDifficulty(difficulty: number): void;

    getExportSettings(): WorldExportSettings;

    setExportSettings(exportSettings: WorldExportSettings): void;

    getHistory(): HistoryEntry[];

    addHistoryEntry(key: number, ...args: Serializable[]): void;

    getMetadata(): Map<string, any>;

    setMetadata(metadata: Map<string, any>): void;

    getBorderSettings(): BorderSettings;

    getMergedWith(): File;

    setMergedWith(mergedWith: File): void;

    getDataPacks(): File[];

    setDataPacks(dataPacks: File[]): void;

    getAttribute<T>(key: AttributeKey<T>): Optional<T>;

    setAttribute<T>(key: AttributeKey<T>, value: T): void;

    transform(transform: CoordinateTransform, progressReceiver: ProgressReceiver): void;

    transform(anchor: Anchor, transform: CoordinateTransform, progressReceiver: ProgressReceiver): void;

    clearLayerData(layer: Layer): void;

    measureSize(): number;

    save(out: ZipOutputStream): void;


}
